README.md
# Picky ☜

Yet another React select list.

[![Build Status](https://travis-ci.org/Aidurber/react-picky.svg?branch=master)](https://travis-ci.org/Aidurber/react-picky)
[![codecov](https://codecov.io/gh/Aidurber/react-picky/branch/master/graph/badge.svg)](https://codecov.io/gh/Aidurber/react-picky)
[![license](https://img.shields.io/github/license/aidurber/react-picky.svg)]()
[![npm version](https://badge.fury.io/js/react-picky.svg)](https://badge.fury.io/js/react-picky)
[![gzip size](https://badgen.net/bundlephobia/minzip/react-picky)]()

## Motivation

When dealing with medium+ length select lists, especially multi-select lists. The common approach is to use tags e.g.

![Tag List](https://raw.githubusercontent.com/aidurber/react-picky/master/assets/readme-tag-list.JPG)

Source: [React-Select by Jed Watson](https://github.com/JedWatson/react-select)

This approach is fine for smaller lists. When you have options for 20, 30, 100+ options that the use can select, it becomes unmanigable.

For example you have a internal staff mailing list. You want to mail all staff in a department (30 employees). You select all. That would be 30 tags taking unneccessary space.

This is a multiselect with checkboxes, a select all option, and a filter. Along a similar vein as [David Stutz's Bootstrap Multiselect](http://davidstutz.github.io/bootstrap-multiselect/). There is a port of that library to react with [Skratchdot's React Bootstrap Multiselect](https://github.com/skratchdot/react-bootstrap-multiselect). However I don't want a dependency on jQuery. If you are already using jQuery that might be an alternative for you.

If you like the tag list like [React-Select](https://github.com/JedWatson/react-select), then that would be a great option for you. It's a really great, well-tested library. Give it a look.

You can also achieve the same result with a great deal of flexibility using [Paypal's Downshift](https://github.com/paypal/downshift#usage).

### More on React Select

React Select V2 is pretty incredible, you can replace and provide custom functionality for almost every element of the select component. With a little work you could rebuild Picky from the components of React Select, take a look at [React Select: Experimental](https://react-select.com/advanced#experimental), the Popout example is close to what Picky does. You could even build a date picker in React Select...**HOW COOL IS THAT!**

## What Picky is

Picky provides a medium amount of flexibility, you can custom render: Options, List (useful for creating a virtualized menu), and SelectAll. Any further customisation and it's a little out of scope for Picky. It was built with a common pattern in mind so you can get up and running with little-to-no work. If you need Picky to be more flexible, I'm happy to take a PR if it would benefit the rest of the community.

# Peer Dependencies

```
 "react": "^16.8.0",
 "react-dom": "^16.8.0"
```

# Installation

```
  npm install --save react-picky
  # or
  yarn add react-picky
```

# Screenshots

## Single Select

![Single select](https://raw.githubusercontent.com/aidurber/react-picky/master/assets/readme-single-select.png)

## Multi Select

![Multi select](https://raw.githubusercontent.com/aidurber/react-picky/master/assets/readme-multi-select.jpg)

# Usage

## Basic example

```javascript
import { Picky } from 'react-picky';
import 'react-picky/dist/picky.css'; // Include CSS

<Picky
  id="picky"
  options={[1, 2, 3, 4, 5]}
  value={[]}
  multiple={true}
  includeSelectAll={true}
  includeFilter={true}
  onChange={values => console.log(values)}
  dropdownHeight={600}
/>;
```

# Examples

- [Basic](./examples/basic)
- [Custom Render: Select All](./examples/custom-selectall)
- [Custom Render: List](./examples/custom-list)
- [Custom Render: Option](./examples/custom-option)

## Props

```javascript
Picky.defaultProps = {
  numberDisplayed: 3,
  options: [],
  filterDebounce: 150,
  dropdownHeight: 300,
  onChange: () => {},
  tabIndex: 0,
  keepOpen: true,
  selectAllText: 'Select all',
  selectAllMode: 'default',
};
Picky.propTypes = {
  id: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.string,
    PropTypes.number,
    PropTypes.object,
  ]),
  numberDisplayed: PropTypes.number,
  multiple: PropTypes.bool,
  options: PropTypes.array.isRequired,
  onChange: PropTypes.func.isRequired,
  open: PropTypes.bool,
  includeSelectAll: PropTypes.bool,
  includeFilter: PropTypes.bool,
  filterDebounce: PropTypes.number,
  dropdownHeight: PropTypes.number,
  onFiltered: PropTypes.func,
  onOpen: PropTypes.func,
  onClose: PropTypes.func,
  valueKey: PropTypes.string,
  labelKey: PropTypes.string,
  render: PropTypes.func,
  tabIndex: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  keepOpen: PropTypes.bool,
  manySelectedPlaceholder: PropTypes.string,
  allSelectedPlaceholder: PropTypes.string,
  selectAllText: PropTypes.string,
  renderSelectAll: PropTypes.func,
  defaultFocusFilter: PropTypes.bool,
  className: PropTypes.string,
  renderList: PropTypes.func,
  filterPlaceholder: PropTypes.string,
  disabled: PropTypes.bool,
  getFilterValue: PropTypes.func,
  caseSensitiveFilter: PropTypes.bool,
  buttonProps: PropTypes.object,
  selectAllMode: PropTypes.oneOf(['default', 'filtered']),
  clearFilterOnClose: PropTypes.bool,
  singleSelectPlaceholder: PropTypes.func,
};
```

### Prop descriptions

- `id` - HTML id assigned to input. Also used as a base for other elements inside Picky.
- `placeholder` - Default message when no items are selected.
- `value` - The selected value(s), array if multiple is true. **Not needed if using as an uncontolled component**
- `numberDisplayed` - Then number of selected options displayed until it turns into '(selected count) selected'.
- `multiple` - Set to true for a multiselect, defaults to false.
- `options` - Array of possible options.
- `onChange` - Called whenever selected value(s) have changed. Pass the selected value back into `value`.
- `open` - Can open or close the dropdown manually. Defaults to false.
- `includeSelectAll` - If set to `true` will add a `Select All` checkbox at the top of the list.
- `includeFilter` - If set to `true` will add an input at the top of the dropdown for filtering the results.
- `filterDebounce` - Debounce timeout, used to limit the rate the internal `onFilterChange` gets called. Defaults to 150ms.
- `dropdownHeight` - The height of the dropdown. Defaults to 300px.
- `onFiltered` - Called after a filter has been done with the filtered options.
- `onOpen` - Called after the dropdown has opened.
- `onClose` - Called after the dropdown has closed.
- `valueKey` - If supplying an array of objects as options, this is required. It's used to identify which property on the object is the value.
- `labelKey` - If supplying an array of objects as options, this is required. It's used to identify which property on the object is the label.
- `render` - Used for custom rendering of items in the dropdown. More info below.
- `tabIndex` - Pass tabIndex to component for accessibility. Defaults to 0
- `keepOpen` - Default true. Single selects close automatically when selecting a value unless this is set to true.
- `manySelectedPlaceholder` - Default "%s selected" where %s is the number of items selected. This gets used when the number if items selected is more than the `numberDisplayed` prop and when all options are not selected.
- `allSelectedPlaceholder` - Default "%s selected" where %s is the number of items selected. This gets used when all options are selected.
- `selectAllText` - Default "Select all", use this to override "Select all" text from top of dropdown when included with `includeSelectAll`.
- `renderSelectAll` - Used for rendering a custom select all
- `defaultFocusFilter` - If set to true, will focus the filter by default when opened.
- `renderList` - Render prop for whole list, you can use this to add virtualization/windowing if necessary
- `filterPlaceholder` - Override the filter placeholder. Defaults to 'Filter...'
- `getFilterValue` - Will provide the input value of filter to the picky dropdown, so that if we have a larger list of options then we can only supply the matching options based on this value.
- `caseSensitiveFilter` - If true options will be returned when they match case
- `buttonProps` - Additional props to apply the the button component, useful for supplying class names.
- `selectAllMode` - default: `default`. When the mode is `filtered` the Select All won't be hidden when filtering.
- `clearFilterOnClose` - When set to true filtered options and filtered state will be cleared on close. Defaults to false.
- `filterTermProcessor` - A function that takes a string and returns a string. Useful for trimming and processing a filter term before it filters the options. Default: (term) => term
- `singleSelectPlaceholder` - A function that takes the currently selected value and returns a string. `(val: OptionType) => string`

## Custom rendering

### Items

You can render out custom items for the dropdown.

**Example**

```javascript
<Picky
  id="picky"
  value={this.state.arrayValue}
  options={bigList}
  onChange={this.selectMultipleOption}
  open={false}
  valueKey="id"
  labelKey="name"
  multiple={true}
  includeSelectAll={true}
  includeFilter={true}
  getFilterValue={this.getFilterValue}
  dropdownHeight={600}
  render={({
    style,
    isSelected,
    item,
    selectValue,
    labelKey,
    valueKey,
    multiple,
  }) => {
    return (
      <li
        style={style} // required
        className={isSelected ? 'selected' : ''} // required to indicate is selected
        key={item[valueKey]} // required
        onClick={() => selectValue(item)}
      >
        {/* required to select item */}
        <input type="checkbox" checked={isSelected} readOnly />
        <span style={{ fontSize: '30px' }}>{item[labelKey]}</span>
      </li>
    );
  }}
/>
```

The render callback gets called with the following properties:
style, isSelected, item, labelKey, valueKey, selectValue, multiple

- `isSelected` - boolean - true if item is selected. Use this for styling accordingly.
- `item` - object | number | string - The item to render.
- `labelKey` - Used to get the label if item is an object
- `valueKey` - Used to get the value if item is an object, good for keys.
- `selectValue` - function(item) - Selects the item on click
- `multiple` - boolean - Indicates if is a multiselect.

**Note**

- If your rendered item affects the height of the item in anyway. Supply `itemHeight` to Picky.
- If you wish to show a radio button or a checkbox, be sure to add `readOnly` prop to the input.

### Select All

```javascript
<Picky
  // ...
  renderSelectAll={({
    filtered,
    tabIndex,
    allSelected,
    toggleSelectAll,
    multiple,
  }) => {
    // Don't show if single select or items have been filtered.
    if (multiple && !filtered) {
      return (
        <div
          tabIndex={tabIndex}
          role="option"
          className={allSelected ? 'option selected' : 'option'}
          onClick={toggleSelectAll}
          onKeyPress={toggleSelectAll}
        >
          <h1>SELECT ALL</h1>
        </div>
      );
    }
  }}
/>
```

Gets called with the following properties:

- `filtered`: boolean - true if items have been filtered.
- `allSelected`: boolean true if all items are selected.
- `toggleSelectAll`: function selects or deselects all items.
- `tabIndex`: number used for specifying tab index.
- `multiple`: boolean true if multiselect.

### Render List

```javascript
<Picky
  id="picky"
  value={this.state.arrayValue}
  options={bigList}
  onChange={this.selectMultipleOption}
  open={true}
  valueKey="id"
  labelKey="name"
  multiple={true}
  includeSelectAll={true}
  includeFilter={true}
  dropdownHeight={600}
  manySelectedPlaceholder={dynamicPlaceholder}
  defaultFocusFilter={true}
  renderList={({ items, selected, multiple, selectValue, getIsSelected }) =>
    items.map(item => (
      <li key={item.id} onClick={() => selectValue(item)}>
        {getIsSelected(item) ? <strong>{item.name}</strong> : item.name}
      </li>
    ))
  }
/>
```

This is an example of a custom rendered list.

# styled-components support

Support is pretty basic by allowing a `className` prop to `<Picky>`, so as a side effect you can add a custom class to the core Picky for easier style overrides.

**Usage**

```javascript
const Select = styled(Picky)`
  background-color: #ff0000;
  .picky__dropdown,
  .option {
    font-size: 2em;
  }
`;
```

# Migrations

## v4 to v5

> Picky is no longer a default export

**V4**

import Picky from 'react-picky'

**V5**

import { Picky } from 'react-picky'


CHANGELOG.md
## [5.3.2](https://github.com/Aidurber/react-picky/compare/v5.3.1...v5.3.2) (2020-09-03)


### Bug Fixes

* effect used conditionally ([0d1da89](https://github.com/Aidurber/react-picky/commit/0d1da89))
* useRef instead of createRef ([7652668](https://github.com/Aidurber/react-picky/commit/7652668))
* wrapper stealing focus ([e599f8c](https://github.com/Aidurber/react-picky/commit/e599f8c)), closes [#221](https://github.com/Aidurber/react-picky/issues/221)

## [5.3.1](https://github.com/Aidurber/react-picky/compare/v5.3.0...v5.3.1) (2020-09-03)


### Bug Fixes

* hooks violation ([f109d37](https://github.com/Aidurber/react-picky/commit/f109d37))

# [5.3.0](https://github.com/Aidurber/react-picky/compare/v5.2.5...v5.3.0) (2020-08-05)


### Features

* add single singleSelectPlaceholder ([d46cf60](https://github.com/Aidurber/react-picky/commit/d46cf60))

## [5.2.5](https://github.com/Aidurber/react-picky/compare/v5.2.4...v5.2.5) (2020-04-14)


### Bug Fixes

* strict mode compliance ([e940925](https://github.com/Aidurber/react-picky/commit/e940925))

## [5.2.4](https://github.com/Aidurber/react-picky/compare/v5.2.3...v5.2.4) (2020-04-12)


### Bug Fixes

* color contrast of placeholder ([cda22c5](https://github.com/Aidurber/react-picky/commit/cda22c5))

## [5.2.3](https://github.com/Aidurber/react-picky/compare/v5.2.2...v5.2.3) (2020-03-10)

### Bug Fixes

- remove prop-types dependency ([#207](https://github.com/Aidurber/react-picky/issues/207)) ([e61c76f](https://github.com/Aidurber/react-picky/commit/e61c76f))

## [5.2.2](https://github.com/Aidurber/react-picky/compare/v5.2.1...v5.2.2) (2020-03-06)

### Bug Fixes

- export picky props ([c186854](https://github.com/Aidurber/react-picky/commit/c186854))

## [5.2.1](https://github.com/Aidurber/react-picky/compare/v5.2.0...v5.2.1) (2020-02-24)

### Bug Fixes

- placeholder empty class ([418c7bd](https://github.com/Aidurber/react-picky/commit/418c7bd))

# [5.2.0](https://github.com/Aidurber/react-picky/compare/v5.1.0...v5.2.0) (2020-01-16)

### Features

- add filter term processor ([794924b](https://github.com/Aidurber/react-picky/commit/794924b))

# [5.1.0](https://github.com/Aidurber/react-picky/compare/v5.0.2...v5.1.0) (2019-11-15)

### Features

- remove recompose ([c07c1db](https://github.com/Aidurber/react-picky/commit/c07c1db))

## [5.0.2](https://github.com/Aidurber/react-picky/compare/v5.0.1...v5.0.2) (2019-11-13)

### Bug Fixes

- **placeholder:** set "picky\_\_placeholder" with a non-empty value ([63bae73](https://github.com/Aidurber/react-picky/commit/63bae73))

## [5.0.1](https://github.com/Aidurber/react-picky/compare/v5.0.0...v5.0.1) (2019-11-03)

### Bug Fixes

- use correct export ([a0edbdc](https://github.com/Aidurber/react-picky/commit/a0edbdc))

# [5.0.0](https://github.com/Aidurber/react-picky/compare/v4.7.2...v5.0.0) (2019-11-03)

### Bug Fixes

- add default id prop ([5a83b8d](https://github.com/Aidurber/react-picky/commit/5a83b8d))
- add default id prop ([f937117](https://github.com/Aidurber/react-picky/commit/f937117))
- classname warnings ([e6b97d1](https://github.com/Aidurber/react-picky/commit/e6b97d1))
- classname warnings ([e6a1688](https://github.com/Aidurber/react-picky/commit/e6a1688))

### Features

- increase react dep to 16.8 ([96b0234](https://github.com/Aidurber/react-picky/commit/96b0234))
- increase react dep to 16.8 ([339c4f6](https://github.com/Aidurber/react-picky/commit/339c4f6))
- migrate to typescript ([1ac1b6f](https://github.com/Aidurber/react-picky/commit/1ac1b6f))
- migrate to typescript ([8ee7298](https://github.com/Aidurber/react-picky/commit/8ee7298))
- select only filtered ([16c164b](https://github.com/Aidurber/react-picky/commit/16c164b))
- support indeternimate select all ([7f1082d](https://github.com/Aidurber/react-picky/commit/7f1082d))
- support indeternimate select all ([fe30273](https://github.com/Aidurber/react-picky/commit/fe30273))

### BREAKING CHANGES

- Requires Hooks
- Large re-write to TypeScript
- Requires Hooks
- Large re-write to TypeScript

## [4.7.2](https://github.com/Aidurber/react-picky/compare/v4.7.1...v4.7.2) (2019-10-17)

### Bug Fixes

- failed prop type and className warnings ([b93dab9](https://github.com/Aidurber/react-picky/commit/b93dab9))

## [4.7.1](https://github.com/Aidurber/react-picky/compare/v4.7.0...v4.7.1) (2019-10-10)

### Bug Fixes

- better styling for placeholder default ([f52619e](https://github.com/Aidurber/react-picky/commit/f52619e)), closes [#144](https://github.com/Aidurber/react-picky/issues/144)

# [4.7.0](https://github.com/Aidurber/react-picky/compare/v4.6.0...v4.7.0) (2019-10-05)

### Bug Fixes

- clear filter input on close ([247b800](https://github.com/Aidurber/react-picky/commit/247b800))

### Features

- support clearing filter on close ([6e7106e](https://github.com/Aidurber/react-picky/commit/6e7106e))

# [4.6.0](https://github.com/Aidurber/react-picky/compare/v4.5.2...v4.6.0) (2019-10-05)

### Features

- add selectAllMode ([27fafd1](https://github.com/Aidurber/react-picky/commit/27fafd1))

## [4.5.2](https://github.com/Aidurber/react-picky/compare/v4.5.1...v4.5.2) (2019-08-08)

### Bug Fixes

- **Accessibility:** Add ID prop ([2bdcf8d](https://github.com/Aidurber/react-picky/commit/2bdcf8d)), closes [#166](https://github.com/Aidurber/react-picky/issues/166)

## [4.5.1](https://github.com/Aidurber/react-picky/compare/v4.5.0...v4.5.1) (2019-07-08)

### Bug Fixes

- **Accessibility:** Dropdown is tab selectable when hidden ([e14ffd6](https://github.com/Aidurber/react-picky/commit/e14ffd6)), closes [#156](https://github.com/Aidurber/react-picky/issues/156)

# [4.5.0](https://github.com/Aidurber/react-picky/compare/v4.4.5...v4.5.0) (2019-07-08)

### Features

- **Button:** Add buttonProps prop ([56e2878](https://github.com/Aidurber/react-picky/commit/56e2878)), closes [#152](https://github.com/Aidurber/react-picky/issues/152)

## [4.4.5](https://github.com/Aidurber/react-picky/compare/v4.4.4...v4.4.5) (2019-05-20)

### Bug Fixes


src/Picky.js
import React from 'react';
import PropTypes from 'prop-types';
import debounce from './lib/debounce';
import includes from './lib/includes';
import {
  isDataObject,
  hasItem,
  keyExtractor,
  hasItemIndex,
  sortCollection,
  arraysEqual,
} from './lib/utils';
import Placeholder from './Placeholder';
import Filter from './Filter';
import Option from './Option';
import './Picky.css';
import SelectAll from './SelectAll';
import Button from './Button';

class Picky extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      selectedValue: props.value || (props.multiple ? [] : null),
      open: props.open,
      filtered: false,
      filteredOptions: [],
      allSelected: false,
    };
    this.toggleDropDown = this.toggleDropDown.bind(this);
    this.toggleSelectAll = this.toggleSelectAll.bind(this);
    this.onFilterChange = this.onFilterChange.bind(this);
    this.selectValue = this.selectValue.bind(this);
    this.allSelected = this.allSelected.bind(this);
    this.handleOutsideClick = this.handleOutsideClick.bind(this);
    this.isItemSelected = this.isItemSelected.bind(this);
    this.focusFilterInput = this.focusFilterInput.bind(this);
    this.getValue = this.getValue.bind(this);
  }
  UNSAFE_componentWillMount() {
    this.setState({
      allSelected: this.allSelected(),
    });
  }

  componentDidMount() {
    this.focusFilterInput(this.state.open);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleOutsideClick, false);
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (
      this.props.options !== nextProps.options ||
      this.props.value !== nextProps.value
    ) {
      let valuesEqual = Array.isArray(nextProps.value)
        ? arraysEqual(nextProps.value, this.props.value)
        : nextProps.value === this.props.value;

      let optsEqual = arraysEqual(nextProps.options, this.props.options);

      this.setState({
        allSelected: !(valuesEqual && optsEqual)
          ? this.allSelected(nextProps.value, nextProps.options)
          : this.allSelected(),
      });
    }
  }

  selectValue(val) {
    const valueLookup = this.props.value;
    if (this.props.multiple && Array.isArray(valueLookup)) {
      const itemIndex = hasItemIndex(
        valueLookup,
        val,
        this.props.valueKey,
        this.props.labelKey
      );

      let selectedValue = [];
      if (itemIndex > -1) {
        selectedValue = [
          ...valueLookup.slice(0, itemIndex),
          ...valueLookup.slice(itemIndex + 1),
        ];
      } else {
        selectedValue = [...this.props.value, val];
      }
      this.setState(
        {
          allSelected: this.allSelected(selectedValue),
        },
        () => {
          this.props.onChange(selectedValue);
        }
      );
    } else {
      this.props.onChange(val);
    }
  }
  /**
   * Get the value of a given option or value safely
   *
   * @param {*} option
   * @returns
   * @memberof Picky
   */
  getValue(option) {
    return typeof this.props.valueKey !== 'undefined'
      ? option[this.props.valueKey]
      : option;
  }
  /**
   * Determine whether all items are selected
   *
   * @returns {Boolean}
   * @memberof Picky
   */
  allSelected(overrideSelected, overrideOptions) {
    const { value, options } = this.props;
    const selectedValue = overrideSelected || value;
    const selectedOptions = overrideOptions || options;

    // If there are no options we are getting a false positive for all items being selected
    if (selectedOptions && selectedOptions.length === 0) {
      return false;
    }
    let copiedOptions = selectedOptions.map(this.getValue);
    let copiedValues = Array.isArray(selectedValue)
      ? selectedValue.map(this.getValue)
      : [];

    return arraysEqual(
      sortCollection(copiedValues),
      sortCollection(copiedOptions)
    );
  }
  /**
   * Toggles select all
   *
   * @memberof Picky
   */
  toggleSelectAll() {
    if (this.props.disabled) return;
    this.setState(
      state => {
        return {
          ...state,
          allSelected: !this.state.allSelected,
        };
      },
      () => {
        if (!this.state.allSelected) {
          this.props.onChange([]);
        } else {
          this.props.onChange(this.props.options);
        }
      }
    );
  }

  isItemSelected(item) {
    return hasItem(
      this.props.value,
      item,
      this.props.valueKey,
      this.props.labelKey
    );
  }

  renderOptions() {
    const items = this.state.filtered
      ? this.state.filteredOptions
      : this.props.options;

    const {
      labelKey,
      valueKey,
      multiple,
      render,
      tabIndex,
      renderList,
      disabled,
    } = this.props;
    if (renderList) {
      return renderList({
        items,
        selected: this.props.value,
        multiple,
        tabIndex,
        getIsSelected: this.isItemSelected,
        selectValue: this.selectValue,
        disabled,
      });
    }
    return items.map((item, index) => {
      // Create a key based on the options value
      const key = keyExtractor(item, valueKey, labelKey);

      const isSelected = this.isItemSelected(item);
      // If render prop supplied for items call that.
      if (typeof render === 'function') {
        return render({
          index,
          item,
          isSelected,
          selectValue: this.selectValue,
          labelKey: labelKey,
          valueKey: valueKey,
          multiple: multiple,
          disabled,
        });
      } else {
        // Render a simple option
        return (
          <Option
            key={key}
            item={item}
            isSelected={isSelected}
            selectValue={this.selectValue}
            labelKey={labelKey}
            valueKey={valueKey}
            multiple={multiple}
            tabIndex={tabIndex}
            disabled={disabled}
            id={this.props.id + '-option-' + index}
          />
        );
      }
    });
  }
  /**
   * Called when Filter term changes. Sets filteredOptions and filtered state.
   *
   * @param {any} term
   * @returns
   * @memberof Picky
   */
  onFilterChange(term) {
    /**
     * getFilterValue function will provide the input value of filter to the picky dropdown, so that if we have a larger list of options then we can only supply the matching options based on this value
     */
    if (this.props.getFilterValue) {
      this.props.getFilterValue(term);
    }
    if (!term.trim()) {
      return this.setState({
        filtered: false,
        filteredOptions: [],
      });
    }
    const isObject = isDataObject(
      this.props.options && this.props.options[0],
      this.props.valueKey,
      this.props.labelKey
    );
    const filteredOptions = this.props.options.filter(option => {
      if (isObject) {
        return includes(
          option[this.props.labelKey],
          term,
          this.props.caseSensitiveFilter
        );
      }
      return includes(option, term, this.props.caseSensitiveFilter);
    });
    this.setState(
      {
        filtered: true,
        filteredOptions,
      },
      () => {
        if (this.props.onFiltered) {
          this.props.onFiltered(filteredOptions);
        }
      }
    );
  }
  /**
   *
   * Called by a click event listener. Used to determine any clicks that occur outside of the component.
   * @param {MouseEvent} e
   * @returns
   * @memberof Picky
   */
  handleOutsideClick(e) {
    // If keep open then don't toggle dropdown
    // If radio and not keepOpen then auto close it on selecting a value
    // If radio and click to the filter input then don't toggle dropdown
    const keepOpen = this.props.keepOpen || this.props.multiple;
    if (this.node && this.node.contains(e.target) && keepOpen) {
      return;
    }
    if (
      this.filter &&
      this.filter.filterInput &&
      this.filter.filterInput.contains(e.target)
    ) {
      return;
    }
    this.toggleDropDown();
  }

  focusFilterInput(isOpen) {
    if (!this.filter || !this.filter.filterInput) return;
    if (isOpen && this.props.defaultFocusFilter) {
      this.filter.filterInput.focus();
    }
    if (!isOpen && this.props.clearFilterOnClose === true) {
      this.filter.filterInput.value = '';
    }
  }
  /**
   * Toggle state of dropdown
   *
   * @memberof Picky
   */
  toggleDropDown() {
    if (!this.state.open) {
      // Add event listener to listen for clicks to determine if click occured outside the component or not
      document.addEventListener('click', this.handleOutsideClick, false);
    } else {
      // Remove
      document.removeEventListener('click', this.handleOutsideClick, false);
    }

    this.setState(
      state => {
        return {
          ...state,
          // Toggle open state
          open: !state.open,
          filtered: this.props.clearFilterOnClose ? false : state.filtered,
          filteredOptions: this.props.clearFilterOnClose
            ? []
            : state.filteredOptions,
        };
      },
      () => {
        const isOpen = this.state.open;
        // Prop callbacks
        this.focusFilterInput(isOpen);
        if (isOpen && this.props.onOpen) {
          this.props.onOpen();
        } else if (!isOpen && this.props.onClose) {
          this.props.onClose();
        }
      }
    );
  }

  get filterDebounce() {
    const { filterDebounce } = this.props;
    return filterDebounce > 0
      ? debounce(this.onFilterChange, filterDebounce)
      : this.onFilterChange;
  }

  get showSelectAll() {
    const { renderSelectAll, multiple, includeSelectAll } = this.props;
    return (
      !renderSelectAll &&
      includeSelectAll &&
      multiple &&
      ((this.props.selectAllMode === 'default' && !this.state.filtered) ||
        this.props.selectAllMode === 'filtered')
    );
  }
  render() {
    const {
      className,
      placeholder,
      value,
      multiple,
      numberDisplayed,
      includeFilter,
      valueKey,
      labelKey,
      tabIndex,
      dropdownHeight,
      renderSelectAll,
      filterPlaceholder,
      disabled,
      buttonProps,
    } = this.props;
    const { open } = this.state;
    let ariaOwns = '';
    if (open) {
      ariaOwns += this.props.id + '-list';
    }
    const buttonId = `${this.props.id}__button`;
    const dropdownStyle = { maxHeight: dropdownHeight, overflowY: 'scroll' };
    return (
      <div
        ref={node => {
          this.node = node;
        }}
        className={['picky', className].join(' ')}
        id={this.props.id}
        role="combobox"
        aria-controls={buttonId}
        aria-expanded={open}
        aria-haspopup={open}
        aria-owns={ariaOwns}
        tabIndex={tabIndex}
      >
        <Button
          id={`${this.props.id}__button`}
          disabled={disabled}
          onClick={this.toggleDropDown}
          {...buttonProps}
        >
          <Placeholder
            allSelected={this.state.allSelected}
            placeholder={placeholder}
            manySelectedPlaceholder={this.props.manySelectedPlaceholder}
            allSelectedPlaceholder={this.props.allSelectedPlaceholder}
            value={value}
            multiple={multiple}
            numberDisplayed={numberDisplayed}
            valueKey={valueKey}
            labelKey={labelKey}
            data-testid="placeholder-component"
          />
        </Button>
        <div
          className="picky__dropdown"
          id={this.props.id + '-list'}
          aria-hidden={!open}
          hidden={!open}
          style={open ? dropdownStyle : { visibility: 'hidden' }}
        >
          {includeFilter && (
            <Filter
              ref={filter => (this.filter = filter)}
              placeholder={filterPlaceholder}
              onFilterChange={this.filterDebounce}
            />
          )}
          {renderSelectAll ? (
            renderSelectAll({
              filtered: this.state.filtered,
              allSelected: this.state.allSelected,
              toggleSelectAll: this.toggleSelectAll,
              tabIndex,
              multiple,
              disabled,
            })
          ) : (
            <SelectAll
              visible={this.showSelectAll}
              tabIndex={tabIndex}
              disabled={disabled}
              allSelected={this.state.allSelected}
              id={this.props.id}
              selectAllText={this.props.selectAllText}
              toggleSelectAll={this.toggleSelectAll}
            />
          )}
          {open && <div data-testid="dropdown">{this.renderOptions()}</div>}
        </div>
      </div>
    );
  }
}

Picky.defaultProps = {
  id: 'picky',
  numberDisplayed: 3,
  options: [],
  filterDebounce: 150,
  dropdownHeight: 300,
  onChange: () => {},
  tabIndex: 0,
  keepOpen: true,
  selectAllText: 'Select all',
  selectAllMode: 'default',
};
Picky.propTypes = {
  id: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.string,
    PropTypes.number,
    PropTypes.object,
  ]),
  numberDisplayed: PropTypes.number,
  multiple: PropTypes.bool,
  options: PropTypes.array.isRequired,
  onChange: PropTypes.func.isRequired,
  open: PropTypes.bool,
  includeSelectAll: PropTypes.bool,
  includeFilter: PropTypes.bool,
  filterDebounce: PropTypes.number,
  dropdownHeight: PropTypes.number,
  onFiltered: PropTypes.func,
  onOpen: PropTypes.func,
  onClose: PropTypes.func,
  valueKey: PropTypes.string,
  labelKey: PropTypes.string,
  render: PropTypes.func,
  tabIndex: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  keepOpen: PropTypes.bool,
  manySelectedPlaceholder: PropTypes.string,
  allSelectedPlaceholder: PropTypes.string,
  selectAllText: PropTypes.string,
  renderSelectAll: PropTypes.func,
  defaultFocusFilter: PropTypes.bool,
  className: PropTypes.string,
  renderList: PropTypes.func,
  filterPlaceholder: PropTypes.string,
  disabled: PropTypes.bool,
  getFilterValue: PropTypes.func,
  caseSensitiveFilter: PropTypes.bool,
  buttonProps: PropTypes.object,
  selectAllMode: PropTypes.oneOf(['default', 'filtered']),
  clearFilterOnClose: PropTypes.bool,
};

export default Picky;
